Flume 2591 datasetsink 2.0

joey opened the pull request.

Commit:
FLUME-2591. Added support for parquet, entity parsing, and a failure policy.

* Added support for having transactions that span calls to process()
* Added support for pluggable entity parsing
* Moved the existing avro parsing code into AvroParser
* Added support for pluggable failure policy
* Added a RetryPolicy that mimics the old behavior
* Added a SavePolicy that saves failed events to a Kite dataset
* Added minimal testing of the new SavePolicy

Still todo:

* Add an explicit test to check for a transaction that spans process()
* Adding more comprhensive tests, especially in the face of different
  expected errors.

Commit:
FLUME-2591. Added tests for the DatasetSink 2.0

* Added tests for setting commit-on-batch to true and false
* Added tests for rollWriter()
* Set keep-alive for the channel speeding up tests ~14x
* Updated shouldRoll() to check >= rollIntervalS to fix
  testTimedFileRoll(). This makes more sense anyway as we
  should have to wait the rollIntervalS+1 seconds to roll.

Commit:
FLUME-2591. Refactored FailurePolicy and EntityParser to be created by factories.

* This allows setting a custom policy or parser class.
* Also added tests on the behavior of DatasetWriter errors.

Commit:
FLUME-2591. Moved parser and policy classes into their own packages.

Commit:
FLUME-2591. Documented new configuration parameters.

Commit:
FLUME-2591. Added tests for improperly serialized records.

* I tested to make sure that even if the schema in the header is
  compatible, if the event was serialized with an incompatible
  schema then it will invoke the FailurePolicy

rdblue commented on flume-ng-doc/sphinx/FlumeUserGuide.rst:
```diff
@@ -2166,10 +2166,29 @@ Property Name            Default  Description
+kite.commitOnBatch       true     If true, the file will be synced and the Flume transaction
+                                  will be commited on each batch of ``kite.batchSize`` records.
+                                  Parquet doesn't support sync, so this setting will be ignored
+                                  and the behavior will be the same as if this was ``false``.
```
Why default to true and ignore this for Parquet? I think it makes sense to fail setup if the setting is incorrect, rather than ignoring the user's choice. Then don't add a default so that the behavior when unset is chosen by the format, false for Parquet and true for Avro.

joey replied:
The only problem with that is we can't validate the parameter in the config because the target dataset may not yet exist. That means we have to delay until the sink has been started to detect the configuration error. I could change the parameter to kite.avro.commitOnBatch. The only issue is how to what effect this should, if any, have on HBase datasets or future dataset implementations we may add in the future.

rdblue replied:
I like the `kite.avro.commitOnBatch` solution. We can handle HBase when it comes up.

joey replied:
Done.

rdblue commented on flume-ng-doc/sphinx/FlumeUserGuide.rst:
```diff
@@ -2166,10 +2166,29 @@ Property Name            Default  Description
+kite.entityParser        avro     Parser that turns Flume ``Events`` into Kite entities.
+                                  Valid values are ``avro`` and the fully-qualified class name
+                                  of an implementation of the ``EntityParser.Builder`` interface.
+kite.failurePolicy       retry    Policy that handles non-recoverable erros such as a missing
```
Typo: "erros" => "errors"

joey replied:
Done. (not really, just a note to myself :) )

rdblue commented on flume-ng-doc/sphinx/FlumeUserGuide.rst:
```diff
@@ -2166,10 +2166,29 @@ Property Name            Default  Description
+                                  raw ``Event`` to the ``kite.error.dataset.uri`` dataset, and the
+                                  fully-qualified class name of an implementation of the
+                                  ``FailurePolicy.Builder`` interface.
+kite.error.dataset.uri   --       URI of the dataset failed events are saved when
```
"... _where_ failed events are saved..."?

joey replied:
yup, that's what I meant to write :)

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -1,45 +1,36 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
```
Nit: unnecessary change.

joey replied:
Sorry, that was my auto-formatter going over-board.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -48,144 +39,157 @@
- * {@code kite.dataset.name}.
+ * Sink that writes events to a Kite Dataset. This sink will parse the body of
+ * each incoming event and store the resulting entity in a Kite Dataset. It
+ * determines datasetUri Dataset by opening a dataset URI
```
This doesn't make sense with "datasetUri" in place of "target". It looks like an automatic refactor change from renaming a variable.

joey replied:
Yup, it was at that.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -48,144 +39,157 @@
+ * each incoming event and store the resulting entity in a Kite Dataset. It
+ * determines datasetUri Dataset by opening a dataset URI
+ * {@code kite.dataset.uri} or opening a repository URI, {@code kite.repo.uri},
+ * and loading a Dataset by name, {@code kite.dataset.name}.
```
This should be "namespace and name" now that namespace has been added.

joey replied:
Good catch.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -48,144 +39,157 @@
 
-    String datasetURI = context.getString(
-      DatasetSinkConstants.CONFIG_KITE_DATASET_URI);
+    // Get the dataset URI and name from the context
```
URI and name?

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -48,144 +39,157 @@
+      String repositoryURI = context.getString(CONFIG_KITE_REPO_URI);
+      Preconditions.checkNotNull(repositoryURI, "Repository URI is missing."
+          + " You must set either " + CONFIG_KITE_DATASET_URI + " or both"
+          + " " + CONFIG_KITE_REPO_URI + " and " + CONFIG_KITE_DATASET_NAME);
```
I think this should only encourage the user to set the dataset URI because the other settings are deprecated.

joey replied:
So just say "No dataset configured, kite.dataset.uri is required."

rdblue replied:
Yes.

joey replied:
Done.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -48,144 +39,157 @@
+
+  // Factories
+  private EntityParserFactory parserFactory = new EntityParserFactory();
+  private FailurePolicyFactory policyFactory = new FailurePolicyFactory();
```
Could these be static, or can they not be shared?

joey replied:
Good call, I'll make the static and final.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -48,144 +39,157 @@
   @Override
   public synchronized void start() {
-    this.lastRolledMs = System.currentTimeMillis();
+    lastRolledMs = System.currentTimeMillis();
```
Why did you remove the reference to `this`? I consider it best practice to call out when you are setting an instance variable.

joey replied:
I've never liked that style, but I can keep it to be more consistent.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -219,52 +249,47 @@ public synchronized void stop() {
-      transaction.begin();
+      // TODO: Can the same Sink object be called from multiple threads
+      if (transaction == null) {
+        transaction = channel.getTransaction();
```
Should also use `this.transation = ...`

joey replied:
Done.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -219,52 +249,47 @@ public synchronized void stop() {
 
-      transaction = channel.getTransaction();
-      transaction.begin();
+      // TODO: Can the same Sink object be called from multiple threads
```
There is usually one driver thread for the sink: https://github.com/apache/flume/blob/trunk/flume-ng-core/src/main/java/org/apache/flume/SinkRunner.java#L135

rdblue replied:
There's also a note about thread-safety in the [Sink API](https://github.com/apache/flume/blob/trunk/flume-ng-core/src/main/java/org/apache/flume/Sink.java#L41):

> While the `Sink#process()` call is guaranteed to only be accessed by a single thread, other calls may be concurrently accessed and should thus be protected.

joey replied:
Ok, so that means that `stop()` can be called while `process()` is still happening so I need a way to make sure that `process()` finishes before committing the transaction in stop(). Correct?

rdblue replied:
It appears that way, but that sounds crazy (stopping in another thread when one might be running it?) and should be done by signalling the SinkRunner to stop (after process finishes), then stopping/joining the thread, and finally by stopping the processor (which could be done by SinkRunner when it is notified). Let's ask for clarification from Hari to make sure.

rdblue replied:
Actually, we can just look at the implementation: https://github.com/apache/flume/blob/trunk/flume-ng-core/src/main/java/org/apache/flume/SinkRunner.java#L96

What I described looks like what happens.

joey replied:
Is it safe to assume that behavior will stay the same? I agree it's what makes sense, but how defensive do we want to be given the API spec says that calls to other methods could be concurrent.

At the very least I should put in a descriptive comment on the reasoning so that if something changed down the line it'd be easier to find.

joey replied:
I'll just put in a comment mentioning the API spec and implementation of SinkRunner to keep the code simpler.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -219,52 +249,47 @@ public synchronized void stop() {
+      if (commitOnBatch) {
+        // End the batch with the failure newPolicy first. If this fails, we'll
+        // end up rolling back the transaction, so we don't need to do the sync
+        policy.endBatch();
```
Where is startBatch() called? Is there no startBatch()?

joey replied:
I didn't see a need for one in the current implementations of the FailurePolicy, but it probably makes sense to add it for consistency.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -48,144 +39,157 @@
+   * A class implementing a failure newPolicy for events that had a
+ non-recoverable error during processing.
+   */
+  private FailurePolicy policy = null;
```
It would be more clear in the code that uses this if it were named `failurePolicy`.

joey replied:
Done.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -219,52 +249,47 @@ public synchronized void stop() {
-      this.writer = newWriter(login, target);
-      this.lastRolledMs = System.currentTimeMillis();
-      LOG.info("Rolled writer for " + getName());
+    if (shouldRoll()) {
```
Shouldn't this check happen before the possible call to newWriter?

joey replied:
So, the only purpose of the above check that creates a new writer if the writer is null is to handle the case where the dataset doesn't exist when the sink starts, otherwise it will reset the rolled time and and not try to open a new writer until the roll time hits again.

I think the cleanest thing for me to do is to not reset the last rolled time unless a new writer was successfully created. That would remove the extra check of a null writer above. The only potential downside is that you'd probably end up with a weird looking log message when it logs that the writer was rolled when it starts up.

joey replied:
Ok, I refactored things a bit and I'll a null writer as a condition for rolling the writer since closeWriter() will check in the writer is null before trying to close it and it will only log when the writer is successfully closed.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -285,94 +310,222 @@ public Status process() throws EventDeliveryException {
+   *                                records were persisted.
+   */
+  @VisibleForTesting
+  void rollWriter(boolean closeWriter, boolean createNewWriter)
```
Uses of this method aren't very clear because it [accepts two booleans](http://qt-project.org/wiki/API_Design_Principles#a6396e641a952364172f8879fe3eb4c1). It would be better to break it into separate methods (close and create) that can be invoked separately, or to get rid of at least one argument by renaming it to "rollWriterInternal" with the boolean and adding "closeAndRollWriter" and "rollWithoutClose" or similar methods that are easier to read.

joey replied:
Yeah, I'll see how I can rework this base don the places where I needed to use it. It evolved as a single method as it made the error handling easier.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -285,94 +310,222 @@ public Status process() throws EventDeliveryException {
+    } catch (RuntimeException ex) {
+      throw new EventDeliveryException("Error trying to open a new"
+          + " writer for dataset " + datasetUri + ": "
+          + ex.getLocalizedMessage(), ex);
```
No need to add the message if also passing ex as the cause.

joey replied:
Done.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -285,94 +310,222 @@ public Status process() throws EventDeliveryException {
+        currentTimeMillis - lastRolledMs);
+
+    LOG.debug("Current time: {}, lastRolled: {}, diff: {} sec",
+        new Object[] {currentTimeMillis, lastRolledMs, elapsedTimeSeconds});
```
Nit: unit names are inconsistent: "Ms" or "Millis"

joey replied:
Fixed.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -285,94 +310,222 @@ public Status process() throws EventDeliveryException {
+      writer = null;
+
+      // commit transaction
+      if (transaction != null) {
```
I think it would be more clear to have all of the transaction logic in process, rather than in certain method calls.

joey replied:
I can't have all of it in process as it also needs to be called from stop(). I can put it into methods.

joey replied:
Just to make my comment clearer. I'm putting the logic into three method calls and will never access the transaction directly.

rdblue replied:
That's a good plan.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -285,94 +310,222 @@ public Status process() throws EventDeliveryException {
+              + " dataset " + datasetUri + ": " + ex.getLocalizedMessage(), ex);
+        }
+      }
+      writer = null;
```
What happens if there is a current writer, but close is false? It would be easier to reason about this code if it were broken into idempotent methods called at the right time, like `closeWriter()` that guarantees the writer is closed correctly and the instance variable is null. This is what newWriter already does, I think.

joey replied:
You don't always want to close the writer because if the writer has thrown a RuntimeException then it won't be in the Error state and doing a close will move the temp file and guarantee there is duplicated data.

This would be cleaner with CDK-885 in.

joey replied:
I've split rollWriter into three methods to make it clearer.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -285,94 +310,222 @@ public Status process() throws EventDeliveryException {
+      }
+
+      if (closeWriter) {
+        // Only log if we closed the old writer
```
Why isn't this done above when the writer is closed?

joey replied:
The logging? Yes it could be moved there.

rdblue replied:
Anything within the `if (closeWriter)` check. Is this something that should be done if and only if the writer is closed, or whenever the writer would be closed if it wasn't null.

joey replied:
This was just to avoid spurious logging but it goes away by refactoring rollLog into three methods.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -285,94 +310,222 @@ public Status process() throws EventDeliveryException {
+
+      Throwables.propagateIfInstanceOf(ex, EventDeliveryException.class);
+      throw new EventDeliveryException("Failed to roll writer for " + datasetUri
+          + ": " + ex.getLocalizedMessage(), ex);
```
Also don't need to add the localized message here.

joey replied:
Done.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/NonRecoverableEventException.java:
```diff
@@ -0,0 +1,54 @@
+ */
+public class NonRecoverableEventException extends Exception {
+
+  private static final long serialVersionUID = 3485151222482254285L;
```
I think it is fine if this class isn't Serializable, right?

joey replied:
I copied that from Flume's EventDeliveryException, but it wasn't clear why it had the serialVersionUID since it's also not serializable.

joey replied:
Oh wow. So Throwable implements Serializable which means that every Exception also implements Serializable. So I'm going to leave this here.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/parser/EntityParserFactory.java:
```diff
@@ -0,0 +1,82 @@
+        DEFAULT_ENTITY_PARSER);
+
+    if (parserType.equals(AVRO_ENTITY_PARSER)) {
+      parser = new AvroParser.Builder().build(datasetSchema, config);
```
This seems unnecessary.

joey replied:
What seems unnecessary?

rdblue replied:
Going through a builder to instantiate the AvroParser instead of doing it directly. Not a blocker though.

joey replied:
Ah. That just made it more consistent with the way the pluggable interface
works. That also makes it consistent with EventSerializers and
EventDeserializers which are related concepts.
On Tue, Jan 27, 2015 at 09:10 Ryan Blue notifications@github.com wrote:

> In
> flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/parser/EntityParserFactory.java
> https://github.com/joey/flume/pull/1#discussion_r23624064:
> 
> > +import org.apache.avro.generic.GenericRecord;
> > +import org.apache.flume.Context;
> > +
> > +import static org.apache.flume.sink.kite.DatasetSinkConstants.*;
> > +
> > +
> > +public class EntityParserFactory {
> > +
> > -  public EntityParser<GenericRecord> newParser(Schema datasetSchema, Context config) {
> > -    EntityParser<GenericRecord> parser;
> >   +
> > -    String parserType = config.getString(CONFIG_ENTITY_PARSER,
> > -        DEFAULT_ENTITY_PARSER);
> >   +
> > -    if (parserType.equals(AVRO_ENTITY_PARSER)) {
> > -      parser = new AvroParser.Builder().build(datasetSchema, config);
> 
> Going through a builder to instantiate the AvroParser instead of doing it
> directly. Not a blocker though.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/joey/flume/pull/1/files#r23624064.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/policy/FailurePolicy.java:
```diff
@@ -0,0 +1,79 @@
+   *                                will be rolled back and the batch will be
+   *                                retried.
+   */
+  public void endBatch() throws EventDeliveryException;
```
Isn't it more appropriate to inherit from `Flushable` for this and name it `flush`? I think that is what the API is doing, but `endBatch` doesn't tell me much about what is happening.

rdblue replied:
There should also be a close operation, right? So if we are using per-batch transactions, flush for each batch. Otherwise, close?

joey replied:
If I do that, then I need to make sure that the configured error dataset supports flush since if the user already created it, then it could be a Parquet dataset.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/policy/FailurePolicyFactory.java:
```diff
@@ -0,0 +1,82 @@
+      Class<? extends FailurePolicy.Builder> builderClass;
+      Class c;
+      try {
+        c = Class.forName(policyType);
```
Could this use the `DynConstructors` classes in Kite to make the process a bit easier?

joey replied:
I can take a look. Are DynConstructors in the public Kite API? I see it in org.kitesdk.compat which I don't think has historically been considered part of Kite's public API.

joey replied:
Just to close the loop on this. @rdblue and I spoke offline and agreed we'd do this as future work after Kite had agreed to include DynConstructors in a support part of the API.

Commit:
FLUME-2591. Implemented review feedback.

Major items include:

* Splitting rollWriter into three method calls.
* Putting all of the transaction handling code into methods that are called
  when needed rather than interacting with the transaction object in
  several places.
* Clarifying the lifecycle of the FailurePolicy to make it easier
  to integrate into the durability semantics of the rest of the Sink code.

Commit:
FLUME-2591. Renamed the parser and policy factories for consitency.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -282,97 +303,232 @@ public Status process() throws EventDeliveryException {
+   * @throws EventDeliveryException There was an error starting a new batch
+   *                                with the failure policy.
+   */
+  private void enterTransaction(Channel channel) throws EventDeliveryException {
```
This method guarantees that an open transaction will be available after it is called, so I think the `commitTransaction` and `rollbackTransaction` methods should assume that the transaction is non-null and validate that assumption with a `Precondition`. If there isn't an open transaction but we try to commit or roll back, then something has gone wrong and we want that to be an error. This probably means we can't open the transaction in a block with `finally { rollbackTransaction() }` though.

rdblue replied:
Actually, `rollbackTransaction` is fine to have a null check since it's contract should be to roll back a transaction if there is one. I think just `commitTransaction` should have the precondition check that there is an active transaction.

joey replied:
Yes, that makes sense.

joey replied:
I found a use of commitTransaction where we're not sure that there is an open transaction. When stop() is called, we close the writer and commit the transaction. I can check if the transaction is not null, but that would add a method that checks the transaction directly.

Any other suggestions?

joey replied:
There's another way that the transaction can end up null when we would otherwise call commit. If you have commitOnBatch enabled then you close the transaction and set it to null during the commit. That means that shouldRoll can't check that the transaction is non-null but that means it will fail when it tries to commit the non-null transaction.

I could have two methods, one that checks the transaction and commits if it's open and another that has the precondition. I'd use the conditional version in the commit that happens during a writer roll and `stop()` and use the other method when committing a batch.

Or I can go back to having one method which check if the transaction is open.

rdblue replied:
What about returning a boolean from `commitTransaction` that signals whether or not action was taken? Then you could wrap the call in a Precondition when you know the commit should happen and catch whether or not anything happened for other cases. That would help solve the problem where you commit a batch and need to close the file if the next batch gets rolled back. So something like `committedBatch |= commitTransaction()` in the batch handling code.

joey replied:
I had to have three lines so I can verify that each time we commit we succeed:

```
        boolean committed = commitTransaction();
        Preconditions.checkState(committed,
            "Tried to commit a batch when there was no transaction");
        committedBatch |= committed;
```

But otherwise, this seems to work well.

rdblue replied:
This looks good to me.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -282,97 +303,232 @@ public Status process() throws EventDeliveryException {
+   * @param event The event to write
+   * @throws EventDeliveryException An error occurred trying to write to the
+                                dataset that couldn't or shouldn't be
+                                handled by the failure newPolicy.
```
I think this should be "failure policy" rather than "failure newPolicy"

joey replied:
Done.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -196,75 +199,93 @@ public synchronized void start() {
+
+    try {
+      if (shouldRoll()) {
+        // close the current writer
```
Nit: the comments here are obvious from the method calls.

joey replied:
Done.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -282,97 +303,232 @@ public Status process() throws EventDeliveryException {
+      // This is also no different the the no-op close the writer does if
+      // it's already in the ERROR state. We can just set it to null and a
+      // new one will be created the next call to process.
       this.writer = null;
```
I think this is only the case if we are not committing each batch. If any transaction has been committed for the current writer, then we must close it.

joey replied:
Good catch. I'll close the writer if commitOnBatch is true.

joey replied:
I also check the commitedBatch flag, so we should avoid most duplicate data. We can still end up with a duplicate if you write more data than Avro's internal sync frequency and it has synced data in the batch that is being rolled back. But that's no different than what we guaranteed before.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -196,75 +199,93 @@ public synchronized void start() {
+      }
 
       if (processedEvents == 0) {
         counter.incrementBatchEmptyCount();
```
I think these clean-up tasks should technically be moved outside the try/catch because they shouldn't ever cause a rollback.

joey replied:
Yes, I can move these.

rdblue commented:
This is looking good, I think it is really close.

Commit:
FLUME-2591. Updated with additional review feedback.

* Renamed kite.commitOnBatch to kite.avro.commitOnBatch.
* Modified commitTransaction to return a boolean flag that is true
  when the transaction was actually committed. This lets us set a
  Precondition when committing a batch when kite.avro.commitOnBatch
  is true.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -1,11 +1,10 @@
+ * work for additional information regarding copyright ownership. The ASF
+ * licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
```
Looks like auto-formatting strikes again.

joey replied:
Done.

rdblue commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -196,183 +207,343 @@ public synchronized void start() {
-      this.lastRolledMs = System.currentTimeMillis();
-      LOG.info("Rolled writer for " + getName());
-    }
+      Channel channel = getChannel();
```
Okay, this looks great and I have just one more item. Now that writer creation is handled elsewhere, there is no "create writer if null" at the start of process. That's great, but I think process should have a check that there is a valid writer around this area if that is the assumption in this method.

joey replied:
I'll add a pre-condition before we get the channel. I'll also add a test where I explicitly set the writer to null before process() is called as it should always get created after I handle rolling of the writer above.

Commit:
FLUME-2591. Added check that writer is not null in process()

* Added an explicit test case for that process() handles a
  null writer.
* Also reverted some inadvertant formatting changes.

joey commented:
I think I've updated everything with the latest review feedback. Let me know if you think it's good, and I'll rebase this on upstream trunk and update the patch on JIRA.

tomwhite commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -196,183 +208,346 @@ public synchronized void start() {
+      if (commitOnBatch) {
+        // Sync before commiting. A failure here will result in rolling back
+        // the transaction
+        writer.sync();
```
It might be worth having an option to flush rather than sync. Sync is more expensive, and also isn't supported by some dataset implementations (e.g. HBase).

Also, this should be updated post https://issues.cloudera.org/browse/CDK-863 to test to see if the writer is Flushable or Syncable.

joey replied:
I can note the Flushable/Syncable interface check to the list of follow-up
improvements that require a new kite release.

Do you thank an option like kite.durabilityMethod with values of sync or
flush makes sense?
On Tue, Jan 27, 2015 at 08:32 Tom White notifications@github.com wrote:

> In
> flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java
> https://github.com/joey/flume/pull/1#discussion_r23620845:
> 
> > ## 
> > -      // commit after data has been written and flushed
> > 
> > ## \-      transaction.commit();
> > -      if (processedEvents == 0) {
> > -        counter.incrementBatchEmptyCount();
> > -        return Status.BACKOFF;
> > -      } else if (processedEvents < batchSize) {
> > -        counter.incrementBatchUnderflowCount();
> > -      } else {
> > -        counter.incrementBatchCompleteCount();
> > -      // commit transaction
> > -      if (commitOnBatch) {
> > -        // Sync before commiting. A failure here will result in rolling back
> > -        // the transaction
> > -        writer.sync();
> 
> It might be worth having an option to flush rather than sync. Sync is more
> expensive, and also isn't supported by some dataset implementations (e.g.
> HBase).
> 
> Also, this should be updated post
> https://issues.cloudera.org/browse/CDK-863 to test to see if the writer
> is Flushable or Syncable.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/joey/flume/pull/1/files#r23620845.

joey replied:
Or maybe I should have it be kite.syncIfAvailable that way it's more clear that some Datasets don't support sync?

joey replied:
Any thoughts on the naming of this parameter @rdblue?

rdblue replied:
I think that kite.syncIfAvailable would be okay, and should be compatible with kite.flushIfAvailable (because a sync also flushes). However, this is getting awkward again and I don't like having "maybe do something" options. Ideally, I think a better way to phrase it is `kite.syncable.syncOnBatch` because it signals that it only applies to Syncable datasets.

joey replied:
So how about I go with two parameters:

`kite.flushable.commitOnBatch`
`kite.syncable.syncOnBatch`

The first one controls if we commit on each batch and it will default to true. The second one controls whether we also sync on each batch and also defaults to true. The naming implies what they apply to (`Flushable` and `Syncable` `Datasets` respectively) and it works because `Syncable` implies `Flushable` and it doesn't make sense to commit the transaction if you aren't going to at least flush.

tomwhite commented on flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java:
```diff
@@ -196,183 +208,346 @@ public synchronized void start() {
+        parser = ENTITY_PARSER_FACTORY.newParser(datasetSchema, context);
+      }
+
+      this.reuseEntity = !("parquet".equals(formatName));
```
Compare to Formats.PARQUET

joey replied:
Good idea.
On Tue, Jan 27, 2015 at 08:40 Tom White notifications@github.com wrote:

> In
> flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java
> https://github.com/joey/flume/pull/1#discussion_r23620962:
> 
> > -            }
> > -          });
> >   +
> > -      DatasetDescriptor descriptor = view.getDataset().getDescriptor();
> > -      String formatName = descriptor.getFormat().getName();
> > -      Preconditions.checkArgument(allowedFormats().contains(formatName),
> > -          "Unsupported format: " + formatName);
> >   +
> > -      Schema newSchema = descriptor.getSchema();
> > -      if (datasetSchema == null || !newSchema.equals(datasetSchema)) {
> > -        this.datasetSchema = descriptor.getSchema();
> > -        // dataset schema has changed, create a new parser
> > -        parser = ENTITY_PARSER_FACTORY.newParser(datasetSchema, context);
> > -      }
> >   +
> > -      this.reuseEntity = !("parquet".equals(formatName));
> 
> Compare to Formats.PARQUET
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/joey/flume/pull/1/files#r23620962.

rdblue commented on flume-ng-doc/sphinx/FlumeUserGuide.rst:
```diff
@@ -2166,10 +2166,31 @@ Property Name            Default  Description
                                   (deprecated; use kite.dataset.uri instead)
 kite.batchSize           100      Number of records to process in each batch
 kite.rollInterval        30       Maximum wait time (seconds) before data files are released
+kite.avro.commitOnBatch  true     If true, the writer will be synced and the Flume transaction
```
Okay, one more thing on this option. Because we are moving to `Syncable` in the upstream API, it would be better to name this kite.syncable.commitOnBatch so that we can apply it for any Syncable dataset.

joey replied:
Good idea. But I'll currently only check the Format in the code since we
don't have the interface based approach yet.
On Tue, Jan 27, 2015 at 09:08 Ryan Blue notifications@github.com wrote:

> In flume-ng-doc/sphinx/FlumeUserGuide.rst
> https://github.com/joey/flume/pull/1#discussion_r23623956:
> 
> >  kite.dataset.name        --       Name of the Dataset where records will be written
> >                                    (deprecated; use kite.dataset.uri instead)
> >  kite.batchSize           100      Number of records to process in each batch
> >  kite.rollInterval        30       Maximum wait time (seconds) before data files are released
> > +kite.avro.commitOnBatch  true     If true, the writer will be synced and the Flume transaction
> 
> Okay, one more thing on this option. Because we are moving to Syncable in
> the upstream API, it would be better to name this
> kite.syncable.commitOnBatch so that we can apply it for any Syncable
> dataset.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/joey/flume/pull/1/files#r23623956.

Commit:
FLUME-2591. Implemented another batch of review feedback.

* Made the name of the commitOnBatch parameter more general so it will
  be abel to work with any Syncable Dataset in the future.
* Added an option to turn off sync in favor of flush for weaker
  durability with higher performance.

Commit:
FLUME-2591. Clarified the names of the commitOnBatch and syncOnBatch properties.
